daemons/based/based_callbacks.c

```diff
@@ -356,9 +357,19 @@ process_ping_reply(xmlNode *reply)
 
             if(remote_cib && remote_cib->children) {
                 /* Additional debug */
-                xml_calculate_changes(the_cib, remote_cib);
-                pcmk__xml_log_changes(LOG_INFO, remote_cib);
-                crm_trace("End of differences");
+                int rc = pcmk_rc_ok;
+
+                if (logger_out == NULL) {
+                    rc = pcmk__log_output_new(&logger_out);
+                    CRM_LOG_ASSERT(rc == pcmk_rc_ok);
+                }
+
+                if (rc == pcmk_rc_ok) {
+                    pcmk__output_set_log_level(logger_out, LOG_INFO);
+                    xml_calculate_changes(the_cib, remote_cib);
```

Comment:
xml_calculate_changes() can modify the_cib. not sure if that's important if we can't show the changes

Reply:
Good catch, and maybe. I just traced through it, and the only change it can make to `the_cib` is to set `pcmk__xf_skip` flags. Not sure whether that would actually make a difference in this context or not. It's probably better to leave it out of the if-block, just in case. The XML code is so convoluted, I don't want to make guesses or try to be exhaustive.

Reply:
Side note, we have bigger problems if we don't enter that if-block -- we failed to allocate and initialize the logger object. I put a `CRM_LOG_ASSERT()` there but could just as well put a `CRM_ASSERT()`. A bad return code pretty much means memory error.

Reply:
The places where I've used `CRM_LOG_ASSERT()` so far are places where I didn't think `CRM_CHECK()` fit well because of lack of a good return action. Places where unless we crash the program outright with `CRM_ASSERT()`, we have more work to do before returning. Here, we'd want to free `remote_cib` and possibly sync the CIB.

See also `cib_utils.c:cib_perform_op()`, `crm_diff.c:generate_patch()`, and `crm_shadow.c:main()`.

Reply:
For tools, assert and check are reasonably equivalent. But for daemons, we don't want to exit (especially uncleanly) except in dire circumstances. Even though this case would likely be a memory error, it's still better to try continue the daemon mainloop in case we can at least clean up a bit.

Reply:
Separately, I've been questioning our use of assert/check for memory errors. Those try to dump core for debugging purposes, and out-of-memory doesn't need that. They're better suited to logic problems than environmental conditions. I tend to think library functions in particular should return memory errors up the call chain, but on the other hand, it is really convenient to assume certain functions never return NULL.

Reply:
That makes sense. And AFAICT `CRM_LOG_ASSERT()` is just `CRM_CHECK()` with an empty failure action. The idea was to continue doing everything else that *doesn't* depend on the logging output object (even though we're probably doomed from memory issues).

Reply:
> questioning our use of assert/check for memory errors. Those try to dump core for debugging purposes...

We could start using `assert()` in those cases.

Reply:
`assert()` is intended to be a no-op in production (NDEBUG). I think for out-of-memory, libraries should bubble up the error code to callers, and tools should exit as cleanly as possible. For daemons, we should probably distinguish between situations where we should go back to the mainloop and those where we should just exit as cleanly as possible.

Reply:
In this particular case, I think `CRM_CHECK()` with return makes sense. If we're out of memory, there's no point trying to calculate the XML changes, but returning to the mainloop may be worthwhile. Alternatively we could just try to log the failure and return.

Reply:
The code in the commit as it stands doesn't try to calculate the XML changes if we failed to initialize the logger. It skips that block, calls `free_xml()` and `sync_our_cib()`, and returns.

I still think we should clean up memory at least, so the `CRM_CHECK()` call would be
```
                    CRM_CHECK(rc == pcmk_rc_ok,                              
                              {free_xml(remote_cib); return;});
```

The only practical difference between that and the current commit with `CRM_LOG_ASSERT()` is that we'd skip `sync_our_cib()`.